### PR TITLE
KAFKA-16224: Do not retry committing if topic or partition deleted

### PR DESCRIPTION
**Pull request by cadonna**
Current logic for auto-committing offsets when partitions are revoked will retry continuously when getting UNKNOWN_TOPIC_OR_PARTITION, leading to the member not completing the revocation in time.

This commit considers error UNKNOWN_TOPIC_OR_PARTITION to be fatal in the context of an auto-commit of offsets before a revocation, even though the error is defined as retriable. This ensures that the revocation can finish in time.

*More detailed description of your change,
if necessary. The PR title and PR message become
the squashed commit message, so use a separate
comment to ping reviewers.*

*Summary of testing strategy (including rationale)
for the feature or bug fix. Unit and/or integration
tests are expected for any behaviour change and
system tests should be considered for larger changes.*

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by cadonna** [comment]
@lianetm Could you please have a look at this PR?

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/CommitRequestManager.java
Comment (AndrewJSchofield) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/CommitRequestManager.java]: RetriableException :)

**Review by AndrewJSchofield — APPROVED** [review]
This is a very sensible improvement. Code looks good to me, and there is a test that verifies the updated behaviour. lgtm.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/CommitRequestManager.java
Comment (cadonna) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/CommitRequestManager.java]: Thanks! Fixed!

**Review by lucasbru — APPROVED** [review]
LGTM, thanks! Waiting for @lianetm to take a look before merging

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/CommitRequestManager.java
Comment (lianetm) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/CommitRequestManager.java]: nit: would probably make the log clearer to all if we add the before revocation part `Auto-commit sync **before revocation** failed` (just like you did for the func name, like it)

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/CommitRequestManager.java
Comment (lianetm) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/CommitRequestManager.java]: do we need these extra parenthesis here?

**Comment by lianetm** [comment]
Thanks for the changes @cadonna , this will have a great impact I expect. Just left some minor comments, otherwise LGTM once the build passes.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/CommitRequestManager.java
Comment (lianetm) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/CommitRequestManager.java]: this is failing the build actually

Comment (cadonna) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/CommitRequestManager.java]: Good catch!

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/CommitRequestManager.java
Comment (cadonna) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/CommitRequestManager.java]: Good idea!

Comment (cadonna) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/CommitRequestManager.java]: I added the `before revocation` to all log messages in this method, since it actually applies to all.

**Review by lianetm — APPROVED** [review]
Thanks for the update. LGTM

**Comment by lianetm** [comment]
Oh one more detail from the build logs we missed:
```
/home/jenkins/workspace/Kafka_kafka-pr_PR-15581/clients/src/test/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImplTest.java:542: error: cannot find symbol
[2024-03-25T09:49:39.907Z]         when(commitRequestManager.maybeAutoCommitSyncNow(anyLong())).thenReturn(commitResult);
```

**Comment by cadonna** [comment]
Sorry for that! I do not know why I am so sloppy with this PR. I am going to fix this now.

**Review by lianetm — APPROVED** [review]
Thanks for the update! LGTM

**Comment by cadonna** [comment]
Build failures are unrelated.

**Pull request merged**
